include/drivers/bbram.h

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright (c) 2021 Google Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_BBRAM_H
+#define ZEPHYR_INCLUDE_DRIVERS_BBRAM_H
+
+#include <device.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * API template to check if the BBRAM is invalid.
+ *
+ * @see z_impl_bbram_check_invlaid
+ */
+typedef int (*bbram_api_check_invalid)(const struct device *dev);
+
+/**
+ * API template to check for standby power failure.
+ *
+ * @see z_impl_bbram_check_standby_power
+ */
+typedef int (*bbram_api_check_standby_power)(const struct device *dev);
+
+/**
+ * API template to check for V CC1 power failure.
+ *
+ * @see z_impl_bbram_check_power
+ */
+typedef int (*bbram_api_check_power)(const struct device *dev);
+
+/**
+ * API template to read from BBRAM.
+ *
+ * @see z_impl_bbram_read
+ */
+typedef int (*bbram_api_read)(const struct device *dev, int offset, int size,
+			      uint8_t *data);
+
+/**
+ * API template to write to BBRAM.
+ *
+ * @see z_impl_bbram_write
+ */
+typedef int (*bbram_api_write)(const struct device *dev, int offset, int size,
+			       uint8_t *data);
+
+__subsystem struct bbram_driver_api {
+	bbram_api_check_invalid check_invalid;
+	bbram_api_check_standby_power check_standby_power;
+	bbram_api_check_power check_power;
+	bbram_api_read read;
+	bbram_api_write write;
+};
+
+__syscall int bbram_check_invalid(const struct device *dev);
+
+/**
+ * @brief Check if BBRAM is invalid
+ *
+ * Check if "Invalid Battery-Backed RAM" status is set then reset the status
+ * bit. This may occur as a result to low voltage at the VBAT pin.
+ *
+ * @param dev BBRAM device pointer.
+ * @return 0 if the Battery-Backed RAM data is valid.
+ */
+static inline int z_impl_bbram_check_invalid(const struct device *dev)
+{
+	const struct bbram_driver_api *api =
+		(const struct bbram_driver_api *)dev->api;
+
+	if (!api->check_invalid) {
+		return -ENOTSUP;
+	}
+
+	return api->check_invalid(dev);
+}
+
+__syscall int bbram_check_standby_power(const struct device *dev);
+
+/**
+ * @brief Check for standby (Volt SBY) power failure.
+ *
+ * Check if the V standby power domain is turned on after it was off then reset
+ * the status bit.
+ *
+ * @param dev BBRAM device pointer.
+ * @return 0 if V SBY power domain is in normal operation.
+ */
+static inline int z_impl_bbram_check_standby_power(const struct device *dev)
+{
+	const struct bbram_driver_api *api =
+		(const struct bbram_driver_api *)dev->api;
+
+	if (!api->check_standby_power) {
+		return -ENOTSUP;
+	}
+
+	return api->check_standby_power(dev);
+}
+
+__syscall int bbram_check_power(const struct device *dev);
+
+/**
+ * @brief Check for V CC1 power failure.
+ *
+ * This will return an error if the V CC1 power domain is turned on after it was
+ * off and reset the status bit.
+ *
+ * @param dev BBRAM device pointer.
+ * @return 0 if the V CC1 power domain is in normal operation.
```

Comment:
@return xxx otherwise?

I cannot find anywhere where it says what the return value is when power was lost

Reply:
Update to `-EFAULT`